udf-compiler/src/main/scala/com/nvidia/spark/udf/Plugin.scala

```diff
@@ -50,7 +51,12 @@ case class LogicalPlanRules() extends Rule[LogicalPlan] with Logging {
     // iterating over NamedExpression
     exp match {
       case f: ScalaUDF => // found a ScalaUDF
-        GpuScalaUDFLogical(f).compile(conf.isTestEnabled)
+        // If it implements the RapidsUDF interface, no need to compile it.
+        if (f.function.isInstanceOf[RapidsUDF]) {
```

Comment:
Was this tested?  I don't think this is going to work as it should never match.  `f.function` is always a lambda function that wraps the actual user's class, if any.  This needs to do something like what [`getRapidsUDFInstance` does in `GpuScalaUDF`](https://github.com/NVIDIA/spark-rapids/blob/branch-21.06/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuScalaUDF.scala#L70-L105) to examine the lambda and see if there's a `RapidsUDF` instance within.

Reply:
I set breakpoint at this line with 2 existing test cases: 1. [a simple UDF compiler test](https://github.com/NVIDIA/spark-rapids/blob/branch-21.06/udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala#L73) 2.[URLDecode test](https://github.com/NVIDIA/spark-rapids/blob/branch-21.06/tests/src/test/scala/com/nvidia/spark/rapids/ScalaUDFSuite.scala#L24)

For 1. it returns false. 
For 2. it returns true.

During my try, I firstly used `f.function.getClass.getInterfaces` and found it contains the `RapidsUDF`, but then I found `isInstanceOf` also works for this case.

![image](https://user-images.githubusercontent.com/20476954/119909449-763afe80-bf87-11eb-9fe0-4b142316d54d.png)

Reply:
I stand correct, there are some instances where `f.function` could be the user's class directly.  However there are also cases where it can be wrapped by Spark in a lambda function, see [the corresponding code in Spark's UDFRegistration](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala#L793-L795).

Therefore I still stand by my comment.  This should not rely solely on direct examination of the instance class.  Also from a maintenance perspective, we should not use different logic in the UDF compiler to detect RapidsUDF instances than the code supporting RapidsUDF.  `getRapidsUDFInstance` should be refactored into a method that can be reused by both the UDF compiler and the RAPIDS UDF support code.

Reply:
Hi @jlowe, I have 2 questions for this:
1) How to compose a test case to cover the lambda function case? I tried to put a lambda Function[ , ](implements RapidsUDF) into the register method but it is still able to detect the RapidsUDF and not use serializedLambda.
2) For refactoring, I found it easy to just remove _private_ keyword for `getRapidsUDFInstance` to let it be used all over, but I don't know if it's appropriate.